File: packages/react-native/ios/HotUpdater/Internal/UnzipServiceFactory.swift
import Foundation

/**
 * Factory for creating the appropriate UnzipService based on content encoding
 */
class UnzipServiceFactory {
    /**
     * Creates an UnzipService based on the content encoding header
     * @param contentEncoding The Content-Encoding header from HTTP response (e.g., "br", nil)
     * @return Appropriate UnzipService implementation
     */
    static func createUnzipService(contentEncoding: String?) -> UnzipService {
        guard let encoding = contentEncoding?.lowercased() else {
            NSLog("[UnzipServiceFactory] Using SSZipArchiveUnzipService (default, no Content-Encoding)")
            return SSZipArchiveUnzipService()
        }

        switch encoding {
        case "br":
            NSLog("[UnzipServiceFactory] Using TarBrotliUnzipService for Content-Encoding: br")
            return TarBrotliUnzipService()
        default:
            NSLog("[UnzipServiceFactory] Unknown Content-Encoding: \(encoding), falling back to SSZipArchiveUnzipService")
            return SSZipArchiveUnzipService()
        }
    }
}

File: packages/react-native/android/src/main/java/com/hotupdater/UnzipServiceFactory.kt
package com.hotupdater

import android.util.Log

/**
 * Factory for creating the appropriate UnzipService based on content encoding
 */
object UnzipServiceFactory {
    private const val TAG = "UnzipServiceFactory"

    /**
     * Creates an UnzipService based on the content encoding header
     * @param contentEncoding The Content-Encoding header from HTTP response (e.g., "br", null)
     * @return Appropriate UnzipService implementation
     */
    fun createUnzipService(contentEncoding: String?): UnzipService =
        when (contentEncoding?.lowercase()) {
            "br" -> {
                Log.d(TAG, "Using TarBrotliUnzipService for Content-Encoding: br")
                TarBrotliUnzipService()
            }
            null, "" -> {
                Log.d(TAG, "Using ZipFileUnzipService (default, no Content-Encoding)")
                ZipFileUnzipService()
            }
            else -> {
                Log.w(TAG, "Unknown Content-Encoding: $contentEncoding, falling back to ZipFileUnzipService")
                ZipFileUnzipService()
            }
        }
}
